File: nexus_core/chunker.py
"""
NEXUS — Chunking Engine
========================
Recursive character splitting with section-heading awareness.
Faithful port of the TypeScript chunker from src/lib/chunker.ts.
"""

from __future__ import annotations

import re
from typing import Optional


# Regex for detecting SEC-style section headings
_SECTION_REGEX = re.compile(
    r"^(?:ITEM\s+\d+[A-Z]?\.?|PART\s+[IVX]+|SECTION\s+\d+|"
    r"RISK\s+FACTORS|MD&A|MANAGEMENT.S\s+DISCUSSION|"
    r"EXECUTIVE\s+SUMMARY|CREDIT\s+RISK|MARKET\s+RISK|"
    r"OPERATIONAL\s+RISK|REGULATORY|COMPLIANCE|"
    r"FINANCIAL\s+HIGHLIGHTS|CLIMATE|ESG)",
    re.IGNORECASE | re.MULTILINE,
)


def _split_recursive(
    text: str,
    separators: list[str],
    chunk_size: int,
    overlap: int,
    sep_index: int,
) -> list[str]:
    """Internal recursive splitter — mirrors the TS implementation."""
    if len(text) <= chunk_size:
        return [text]

    # Exhausted all separators — character-level split
    if sep_index >= len(separators):
        result: list[str] = []
        i = 0
        while i < len(text):
            result.append(text[i : i + chunk_size])
            i += chunk_size - overlap
        return result

    sep = separators[sep_index]
    parts = text.split(sep)
    result = []
    current = ""

    for part in parts:
        candidate = (current + sep + part) if current else part

        if len(candidate) > chunk_size and len(current) > 0:
            result.append(current)
            overlap_text = current[-overlap:] if len(current) > overlap else current
            current = overlap_text + sep + part

            # Single part still exceeds — recurse with next separator
            if len(part) > chunk_size:
                sub_parts = _split_recursive(
                    part, separators, chunk_size, overlap, sep_index + 1
                )
                result.extend(sub_parts[:-1])
                current = sub_parts[-1] if sub_parts else ""
        else:
            current = candidate

    if current.strip():
        result.append(current)

    # Re-check for chunks still too long
    final: list[str] = []
    for chunk in result:
        if len(chunk) > chunk_size * 1.5:
            final.extend(
                _split_recursive(chunk, separators, chunk_size, overlap, sep_index + 1)
            )
        else:
            final.append(chunk)

    return final


def chunk_text_recursive(
    text: str,
    chunk_size: int = 800,
    overlap: int = 120,
    min_chunk_size: int = 80,
) -> list[dict]:
    """
    Recursive character text splitter with section-heading detection.

    Tries separators in priority order:
        '\\n\\n', '\\n', '. ', '? ', '! ', '; ', ' ', character-level

    Each result dict contains:
        content      — the chunk text
        chunk_index  — zero-based index
        section      — detected section heading or None
        word_count   — number of words
        char_count   — number of characters

    Args:
        text:           Raw document text to chunk.
        chunk_size:     Maximum chunk size in characters.
        overlap:        Number of characters to overlap between chunks.
        min_chunk_size: Minimum chunk size; smaller chunks are merged forward.

    Returns:
        List of chunk dicts sorted by chunk_index.
    """
    if not text or not text.strip():
        return []

    separators = ["\n\n", "\n", ". ", "? ", "! ", "; ", " ", ""]

    raw_chunks = _split_recursive(text, separators, chunk_size, overlap, 0)

    # Post-process: detect section headings and assign metadata
    chunks: list[dict] = []
    current_section: Optional[str] = None

    for i in range(len(raw_chunks)):
        raw = raw_chunks[i].strip()
        if len(raw) < min_chunk_size and i < len(raw_chunks) - 1:
            # Too small — merge with next chunk
            raw_chunks[i + 1] = raw + "\n\n" + raw_chunks[i + 1]
            continue

        section_match = _SECTION_REGEX.match(raw)
        if section_match:
            current_section = section_match.group(0).upper()

        words = [w for w in raw.split() if w]
        chunks.append(
            {
                "content": raw,
                "chunk_index": len(chunks),
                "section": current_section,
                "word_count": len(words),
                "char_count": len(raw),
            }
        )

    return chunks

File: src/lib/chunker.ts
/**
 * Chunking Engine
 * Recursive character splitting with section-heading awareness
 */

export interface ChunkOutput {
  content: string;
  chunkIndex: number;
  section: string | null;
  wordCount: number;
  charCount: number;
}

/**
 * Recursive character text splitter.
 * Tries separators in priority order: double newline, single newline,
 * sentence boundary, space, then character-level as last resort.
 */
export function chunkTextRecursive(
  text: string,
  options: { chunkSize?: number; overlap?: number; minChunkSize?: number } = {}
): ChunkOutput[] {
  const {
    chunkSize = 800,
    overlap = 120,
    minChunkSize = 80,
  } = options;

  if (!text || text.trim().length === 0) return [];

  const separators = ['\n\n', '\n', '. ', '? ', '! ', '; ', ', ', ' '];

  const rawChunks = splitRecursive(text, separators, chunkSize, overlap, 0);

  // Post-process: detect section headings and assign metadata
  const sectionRegex = /^(?:ITEM\s+\d+[A-Z]?\.?|PART\s+[IVX]+|SECTION\s+\d+|RISK\s+FACTORS|MD&A|MANAGEMENT.S\s+DISCUSSION|EXECUTIVE\s+SUMMARY|CREDIT\s+RISK|MARKET\s+RISK|OPERATIONAL\s+RISK|REGULATORY|COMPLIANCE|FINANCIAL\s+HIGHLIGHTS|CLIMATE|ESG)/i;

  const chunks: ChunkOutput[] = [];
  let currentSection: string | null = null;

  for (let i = 0; i < rawChunks.length; i++) {
    const raw = rawChunks[i].trim();
    if (raw.length < minChunkSize && i < rawChunks.length - 1) {
      // Too small — merge with next chunk
      rawChunks[i + 1] = raw + '\n\n' + rawChunks[i + 1];
      continue;
    }

    const sectionMatch = raw.match(sectionRegex);
    if (sectionMatch) {
      currentSection = sectionMatch[0].toUpperCase();
    }

    const words = raw.split(/\s+/).filter(w => w.length > 0);
    chunks.push({
      content: raw,
      chunkIndex: chunks.length,
      section: currentSection,
      wordCount: words.length,
      charCount: raw.length,
    });
  }

  return chunks;
}

function splitRecursive(
  text: string,
  separators: string[],
  chunkSize: number,
  overlap: number,
  sepIndex: number
): string[] {
  if (text.length <= chunkSize) return [text];

  // If we've exhausted all separators, do character-level split
  if (sepIndex >= separators.length) {
    const result: string[] = [];
    for (let i = 0; i < text.length; i += chunkSize - overlap) {
      result.push(text.slice(i, i + chunkSize));
    }
    return result;
  }

  const sep = separators[sepIndex];
  const parts = text.split(sep);
  const result: string[] = [];
  let current = '';

  for (let i = 0; i < parts.length; i++) {
    const part = parts[i];
    const candidate = current ? current + sep + part : part;

    if (candidate.length > chunkSize && current.length > 0) {
      result.push(current);
      // Keep overlap from end of current chunk
      const overlapText = current.length > overlap ? current.slice(-overlap) : current;
      current = overlapText + sep + part;

      // If even a single part exceeds chunkSize, recurse with next separator
      if (part.length > chunkSize) {
        const subParts = splitRecursive(part, separators, chunkSize, overlap, sepIndex + 1);
        // The last subPart goes into current (it may still accumulate)
        result.push(...subParts.slice(0, -1));
        current = subParts[subParts.length - 1] || '';
      }
    } else {
      current = candidate;
    }
  }

  if (current.trim()) {
    result.push(current);
  }

  // Check if any chunks are still too long — recurse them with next separator
  const final: string[] = [];
  for (const chunk of result) {
    if (chunk.length > chunkSize * 1.5) {
      final.push(...splitRecursive(chunk, separators, chunkSize, overlap, sepIndex + 1));
    } else {
      final.push(chunk);
    }
  }

  return final;
}

/**
 * Legacy paragraph-based chunker (backward compatible).
 * Kept for Demo Mode which uses pre-chunked data.
 */
export function chunkTextParagraph(
  text: string,
  maxChunkSize = 800,
  minChunkSize = 80,
  overlapSize = 60
): ChunkOutput[] {
  if (!text || text.trim().length === 0) return [];

  const paragraphs = text.split(/\n\s*\n/).filter(p => p.trim().length > 0);
  const chunks: ChunkOutput[] = [];
  let currentChunk = '';
  let chunkIndex = 0;
  let currentSection: string | null = null;

  const sectionRegex = /^(?:ITEM\s+\d+[A-Z]?|PART\s+[IVX]+|SECTION\s+\d+|CHAPTER\s+\d+|RISK\s+FACTORS|MD&A)/i;

  for (const para of paragraphs) {
    const trimmed = para.trim();
    if (!trimmed) continue;

    const sectionMatch = trimmed.match(sectionRegex);
    if (sectionMatch) {
      if (currentChunk.trim().length >= minChunkSize) {
        const words = currentChunk.trim().split(/\s+/).filter(w => w.length > 0);
        chunks.push({
          content: currentChunk.trim(),
          chunkIndex,
          section: currentSection,
          wordCount: words.length,
          charCount: currentChunk.trim().length,
        });
        chunkIndex++;
        const overlapWords = currentChunk.trim().split(/\s+/).slice(-overlapSize).join(' ');
        currentChunk = overlapWords + '\n\n' + trimmed;
      } else {
        currentChunk = currentChunk ? currentChunk + '\n\n' + trimmed : trimmed;
      }
      currentSection = sectionMatch[0].toUpperCase();
    } else if (
      currentChunk.length + trimmed.length > maxChunkSize &&
      currentChunk.trim().length >= minChunkSize
    ) {
      const words = currentChunk.trim().split(/\s+/).filter(w => w.length > 0);
      chunks.push({
        content: currentChunk.trim(),
        chunkIndex,
        section: currentSection,
        wordCount: words.length,
        charCount: currentChunk.trim().length,
      });
      chunkIndex++;
      const overlapWords = currentChunk.trim().split(/\s+/).slice(-overlapSize).join(' ');
      currentChunk = overlapWords + '\n\n' + trimmed;
    } else {
      currentChunk = currentChunk ? currentChunk + '\n\n' + trimmed : trimmed;
    }
  }

  if (currentChunk.trim().length > 0) {
    const words = currentChunk.trim().split(/\s+/).filter(w => w.length > 0);
    chunks.push({
      content: currentChunk.trim(),
      chunkIndex,
      section: currentSection,
      wordCount: words.length,
      charCount: currentChunk.trim().length,
    });
  }

  return chunks;
}
